test/builder-api/test.sh

```diff
@@ -1,34 +1,286 @@
+connection_test = false
+pool_size = 8
+EOF
+
+cat << EOF > $dir/config_originsrv.toml
+[datastore]
+host = "127.0.0.1"
+port = $port
+user = "$user"
+database = "test"
+connection_retry_ms = 300
+connection_timeout_sec = 3600
+connection_test = false
+pool_size = 8
+EOF
+
+cat << EOF > $dir/Procfile
+api: $base_dir/target/debug/bldr-api start --path $dir/depot --config $dir/config_api.toml
+router: $base_dir/target/debug/bldr-router start
+jobsrv: $base_dir/support/run-server jobsrv $dir/config_jobsrv.toml
+sessionsrv: $base_dir/support/run-server sessionsrv $dir/config_sessionsrv.toml
+originsrv: $base_dir/support/run-server originsrv $dir/config_originsrv.toml
+EOF
+
+cat << EOF > $dir/bldr.env
+RUST_LOG=debug,postgres=error,habitat_builder_db=error,hyper=error,habitat_builder_router=error,zmq=error,habitat_net=error
+RUST_BACKTRACE=1
+HAB_DOCKER_STUDIO_IMAGE="habitat-docker-registry.bintray.io/studio"
+EOF
+
+# Travis can't run the worker, due to the inability to add the krangschnak user. If we're
+# not running on Travis, add worker config in.
+if [ -z "$TRAVIS" ]; then
+cat << EOF > $dir/config_worker.toml
+auth_token = "bobo"
+bldr_url = "http://localhost:9636"
+auto_publish = true
+data_path = "/tmp"
+
+[github]
+app_private_key = "$key_dir/builder-github-app.pem"
+EOF
+
+  echo "worker: $base_dir/target/debug/bldr-worker start --config $dir/config_worker.toml" >> $dir/Procfile
+fi
+
+# Start all the services up
+if [ "$platform" = "mac" ]; then
+  echo "Running these tests on a mac"
+  env HAB_FUNC_TEST=1 "$base_dir/support/mac/bin/forego" start -f "$dir/Procfile" -e "$dir/bldr.env" > "$dir/services.log" 2>&1 &
+else
+  echo "Running these tests on linux"
+  # env HAB_FUNC_TEST=1 "$base_dir/support/linux/bin/forego" start -f "$dir/Procfile" -e "$dir/bldr.env" > "$dir/services.log" 2>&1 &
+  env HAB_FUNC_TEST=1 "$base_dir/support/linux/bin/forego" start -f "$dir/Procfile" -e "$dir/bldr.env"
+fi
+
 forego_pid=$!
-popd
 
 echo "**** Spinning up the services ****"
-count=0;
-while [ $count -ne 5 ]; do
-  count=0;
-  for svc in builder-originsrv builder-sessionsrv builder-router builder-api builder-jobsrv; do
-    if grep -q "$svc is ready to go" ./services.log; then
-      ((count++))
-    fi
-  done
-  sleep 1
-done
-echo "**** Services ready ****"
+# total=0
+# originsrv=0
+# sessionsrv=0
+# router=0
+# api=0
+# jobsrv=0
+# worker=0
+
+# while [ $total -ne 6 ]; do
+#   if [ ! -f "$dir/services.log" ]; then
+#     continue
+#   fi
+
+#   for svc in originsrv sessionsrv router api jobsrv worker; do
+#     if grep -q "builder-$svc is ready to go" "$dir/services.log"; then
+#       declare "$svc=1"
+#     else
+#       echo "Waiting on $svc"
+#     fi
+#   done
+
+#   total=$((originsrv + sessionsrv + router + api + jobsrv + worker))
+
+#   echo ""
+#   sleep 1
+# done
+# echo "**** All services ready ****"
+
+# Run the tests
+cd "$base_dir/test/builder-api"
+nvm install
+
+npm install
 npm run mocha
 mocha_exit_code=$?
```

Comment:
If the script uses `set -e` (which I think is a good idea), the approach to getting a nonzero exit code will need to be a little different. Here's my suggestion:

```bash
if npm run mocha; then
  echo "All tests passed"
else
  mocha_exit_code=$?
fi
```
and then move the cleanup after this point to a trap on exit:
```bash
on_exit() {
  if pid_is_running "${forego_pid:-}"; then
    echo "**** Stopping services ****"
    sudo kill "$forego_pid"
  fi

  sudo rm -fr "$dir"
  rm -f neurosis*.hart
  exit "${mocha_exit_code:-0}"
}
trap on_exit EXIT
```
Note that this approach doesn't work:
```bash
if ! npm run mocha; then
  mocha_exit_code=$?
fi
```
Since `!` changes `$?`.